File: src/main/java/com/learn/todolist/application/UserRegisterServiceImpl.java
package com.learn.todolist.application;

import com.learn.todolist.domain.User;
import com.learn.todolist.infra.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;

@RequiredArgsConstructor
@Component
public class UserRegisterServiceImpl {
    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;

    public Boolean registerUser(String email, String nickName, String password) {
        User user = User.builder().email(email).nickName(nickName).password(passwordEncoder.encode(password)).build();
        userRepository.save(user);
        return true;
    }

    public Boolean withdrawUser(String email, String password) {
        User user = userRepository.findByEmail(email).orElseThrow(() -> new IllegalArgumentException("User not found"));
        if (!passwordEncoder.matches(password, user.getPassword())) {
            throw new IllegalArgumentException("Password is not correct");
        }
        user.setDeletedAt(LocalDateTime.now());
        userRepository.save(user);
        return true;
    }
}

File: src/main/kotlin/com/todolist/aplication/user/usecase/service/UserRegisterServiceImpl.kt
package com.todolist.aplication.user.usecase.service

import com.todolist.aplication.commonRepository.UserRepository
import com.todolist.domain.user.entity.User
import com.todolist.domain.user.module.UserRegisterService
import io.github.oshai.kotlinlogging.KotlinLogging
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Component
import java.time.LocalDateTime

private val logger = KotlinLogging.logger {}

@Component
class UserRegisterServiceImpl(
    private val userRepository: UserRepository,
    private val passwordEncoder: PasswordEncoder
): UserRegisterService{
    override fun registerUser(email: String, nickName: String, password: String): Boolean {
        logger.info { "User Entity Create" }
        val registerUser = User(
            email = email,
            nickName = nickName,
            password = passwordEncoder.encode(password)
        )

        userRepository.save(registerUser)
        logger.info { "Successful User Signup Welcome $nickName !" }
        return true
    }

    override fun withdrawalUser(email: String, password: String): Boolean {
        logger.info { "Search Withdrawal Target User Entity" }
        val withdrawalUser: User = userRepository.findByEmail(email) ?: throw UsernameNotFoundException("User Not Found")

        logger.info { "Deleted DateTime Update Delete User Entity" }
        withdrawalUser.deletedAt = LocalDateTime.now()
        userRepository.save(withdrawalUser)
        logger.info { "Successful User Withdrawal GoodBye: ${withdrawalUser.nickName}" }
        return true
    }

}